server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,197 @@
+        this.rerouteServiceSupplier = rerouteServiceSupplier;
+        snapshotShardSizes = ImmutableOpenMap.of();
+    }
+
+    public static class SnapshotShard {
+
+        private final Snapshot snapshot;
+        private final IndexId index;
+        private final int shardId;
+
+        public SnapshotShard(Snapshot snapshot, IndexId index, int shardId) {
+            this.snapshot = snapshot;
+            this.index = index;
+            this.shardId = shardId;
+        }
+
+        public Snapshot getSnapshot() {
+            return snapshot;
+        }
+
+        public IndexId getIndex() {
+            return index;
+        }
+
+        public int getShardId() {
+            return shardId;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            SnapshotShard that = (SnapshotShard) o;
+            return shardId == that.shardId &&
+                snapshot.equals(that.snapshot) &&
+                index.equals(that.index);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(snapshot, index, shardId);
+        }
+
+        @Override
+        public String toString() {
+            return "SnapshotShard{" +
+                "snapshot=" + snapshot +
+                ", index=" + index +
+                ", shardId=" + shardId +
+                '}';
+        }
+    }
+
+    @Override
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(snapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> requiredSnapshotShards = new HashSet<>();
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.id());
+                    requiredSnapshotShards.add(snapshotShard);
+                    // check if already populated entry
+                    if (snapshotShardSizes.get(snapshotShard) == null) {
+                        // check if already fetching snapshot info in progress
+                        if (fetchingShards.add(snapshotShard)) {
+                            // TODO: Use a dedicated threadpool here? Use FETCH thread pool?
+                            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I'm not sure which thread-pool to use here. SNAPSHOT wasn't great because I feared allocation could be delayed by long-running backups. ASYNC_FETCH thread pool could turn out to be problematic as it might interfere with recoveries of local primaries (although that threadpool is generously sized).

Reply:
I agree with not using the SNAPSHOT thread pool here. I think we should reuse FETCH_SHARD_STORE or add a new one.